docs: Adding Spark Best Practices with new guidelines (#856)
awslabs/data-on-eks: website/docs/bestpractices/analytics/spark-best-practices.md

Spark applications running on Kubernetes generate high volumes of DNS lookups when executors communicate with external services. This occurs during data ingestion, processing, and when connecting to external databases or shuffle services.

When DNS traffic exceeds 1024 packets per second for a CoreDNS replica, DNS requests will be throttled, resulting in `unknownHostException` errors.

#### Remediation
It is recommended to scale CoreDNS, as your workload scales. Refer to [Scaling CoreDNS](/docs/bestpractices/networking#scaling-coredns) for more details on implementation choices. 

### Reduce Inter AZ Traffic

#### Problem Statement
During the shuffle stage, Spark executors may need to exchange data between them. If the Pods are spread across multiple Availability Zones (AZs), this shuffle operation can turn out to be very expensive, especially on Network I/O front, which will be charged as Inter-AZ Traffic costs.

#### Remediation
For Spark workloads, it is recommended to colocate executors and worker pods in the same AZ. Colocating workloads in the same AZ serves two main purposes:
        * Reduce inter-AZ traffic costs
        * Reduce network latency between executors/Pods

> 

Reviewer: Please fix indentation

#### Consolidation Configuration
While enabling `consolidation` for Spark executor pods can lead to better cluster resource utilization, it's crucial to strike a balance with job performance. Frequent consolidation events can result in slower execution times for Spark jobs, as executors are forced to recompute the shuffle data and RDD blocks.

This impact is particularly noticeable in long-running Spark jobs. To mitigate this, it's essential to carefully tune the consolidation interval.

To save intermediate data computed in Spark Executors, consider one of the following options, which are discussed in `Storage section`:
    * PVC attach
    * External Shuffle storage
    * Executor handling (Related to handling interruptions)
        * `spark.executor.decommission.enabled=true`
        * `spark.storage.decommission.enabled=true`

#### Handling interruptions during Karpenter Consolidation/Spot Termination

> Perform a controlled decommissioning instead of killing an executor. Leverage `TerminationGracePeriod`, be executor aware and save shuffle data before node being decommissioned.

Reviewer: ```suggestion
Perform controlled decommissioning instead of abruptly killing executors when nodes are scheduled for termination. To achieve this: 

     Configure appropriate TerminationGracePeriod values for Spark workloads
     Implement executor-aware termination handling
     Ensure shuffle data is saved before nodes are decommissioned

```

> Spark also provides native configurations, to control the termination behaviour.
>         * Controlling the `executor` interruptions
>             * Config:
>                 * `spark.executor.decommission.enabled`
>                 * `spark.executor.decommission.forceKillTimeout`
>             * This is particularly useful in scenarios where executors might be terminated due to spot instance interruptions or Karpenter consolidation events.
>             * This will allow executor to gracefully shutdown. Executors will stop accepting tasks and would notify driver about decommissioning state.
>         * To control executor’s `BlockManager` behaviour
>             * Config:
>                 * `spark.storage.decommission.enabled`
>                 * `spark.storage.decommission.shuffleBlocks.enabled`
>                 * `spark.storage.decommission.rddBlocks.enabled`
>                 * `spark.storage.decommission.fallbackStorage.path`
> 
> This enables the migration of shuffle and RDD blocks from decommissioning executor to other available executors or to a fallback storage location if configured. This helps in dynamic environment by reducing the need to recompute the shuffle data or RDD blocks, there by improving the job completion times and resource efficiency.

Reviewer: ```suggestion
Spark provides native configurations to control termination behavior:

**Controlling executor interruptions**
* **Configs**:
  * `spark.executor.decommission.enabled`
  * `spark.executor.decommission.forceKillTimeout`

These configurations are particularly useful in scenarios where executors might be terminated due to spot instance interruptions or Karpenter consolidation events. When enabled, executors will gracefully shutdown by stopping task acceptance and notifying the driver about their decommissioning state.

**Controlling executor's BlockManager behavior**
* **Configs**:
  * `spark.storage.decommission.enabled`
  * `spark.storage.decommission.shuffleBlocks.enabled`
  * `spark.storage.decommission.rddBlocks.enabled`
  * `spark.storage.decommission.fallbackStorage.path`

These settings enable the migration of shuffle and RDD blocks from decommissioning executors to other available executors or to a fallback storage location. This approach helps in dynamic environments by reducing the need to recompute shuffle data or RDD blocks, thereby improving job completion times and resource efficiency.
```

## Advanced Scheduling Considerations
### Default Kubernetes Scheduler behaviour.

> Default kubernetes scheduler uses `least allocated` approach. This strategy aims to distribute pods evenly across cluster, which helps in maintaining availability and a balanced resource utilization across all nodes, rather than packing more pods in fewer nodes.

Reviewer: ```suggestion
Default Kubernetes scheduler uses `least allocated` approach. This strategy aims to distribute pods evenly across cluster, which helps in maintaining availability and a balanced resource utilization across all nodes, rather than packing more pods in fewer nodes.
```

`Most allocated` approach on the other hand, aims to favor nodes with most amount of allocated resources, which leads to packing more pods onto nodes that are already heavily allocated. This approach is favourable for spark jobs, as it aims for high utilization on select nodes at pod scheduling time, leading to better consolidation of nodes. You will have to leverage a custom kube-scheduler with this option enabled, or leverage Custom Schedulers purpose built for more advanced orchestration.

### Custom Schedulers

Custom schedulers enhance Kubernetes’ native scheduling capabilities by providing advanced features tailored for batch and high-performance computing workloads. Custom schedulers enhance resource allocation by optimizing bin-packing and offering scheduling tailored to specific application needs. Here are popular custom schedulers for running Spark workloads on Kubernetes.

>     * [Apache Yunikorn](https://yunikorn.apache.org/)
>     * [Volcano](https://volcano.sh/en/)

Reviewer: ```suggestion
* [Apache Yunikorn](https://yunikorn.apache.org/)
* [Volcano](https://volcano.sh/en/)
```

> Advantages of leveraging custom schedulers like Yunikorn.
>         * It provides hierarchical queue system and configurable policies allow for complex resource management.
>         * Gang Scheduling

Reviewer: ```suggestion
#### Advantages of leveraging custom schedulers

* Hierarchical queue system and configurable policies allowing for complex resource management
* Gang scheduling, which ensures all related pods (like Spark executors) start together, preventing resource wastage
* Resource fairness across different tenants and workloads
```

> ### How will Yunikorn and Karpenter work together?
> 
> Karpenter focuses on node provisioning and scaling, while custom schedulers like Yunikorn, brings in application awareness for scheduling.
> 
> Yunikorn manages the advanced application scheduling aspects like queue management, resource fairness, gang scheduling.Once Yunikorn schedules the pods, based on application awareness, Karpenter will then fulfil needs of pending pods, spawned by Yunikorn.

Reviewer: ```suggestion
### How Yunikorn and Karpenter Work Together

Karpenter and Yunikorn complement each other by handling different aspects of workload management in Kubernetes:

* **Karpenter** focuses on node provisioning and scaling, determining when to add or remove nodes based on resource demands.

* **Yunikorn** brings application awareness to scheduling through advanced features like queue management, resource fairness, and gang scheduling.

In a typical workflow, Yunikorn first schedules pods based on application-aware policies and queue priorities. When these pods remain pending due to insufficient cluster resources, Karpenter detects these pending pods and provisions appropriate nodes to accommodate them. This integration ensures both efficient pod placement (Yunikorn) and optimal cluster scaling (Karpenter).

For Spark workloads, this combination is particularly effective: Yunikorn ensures executors are scheduled according to application SLAs and dependencies, while Karpenter ensures the right node types are available to meet those specific requirements.
```

## Storage Best Practices
### Node Storage

> By default, the EBS root volumes of worker nodes is set to 20GB. Spark Executors use local storage for temporary data like shuffle data, intermediate results and temporary files. Default storage of 20GB of root volume attached to worker nodes can be limiting in size and performance. Consider the following options to cater to your performance and storage size requirements.

Reviewer: ```suggestion
By default, the EBS root volumes of worker nodes are set to 20GB. Spark Executors use local storage for temporary data like shuffle data, intermediate results, and temporary files. This default 20GB root volume can be limiting in both size and performance. Consider the following options to address your performance and storage requirements:

```

>         * Expand the root volume capacity.

Reviewer: Any ball park number for this? 100GB etc. Fix indentation please.

        * Configure high-performance storage with better I/O and latency.

>         * Mount additional volumes on worker nodes for temporary data storage.

Reviewer: ```suggestion
* Mount additional volumes on worker nodes for temporary data storage
```

>         * Leverage dynamically provisioned PVCs that can be attached directly to executor pods.

Reviewer: ```suggestion
* Leverage dynamically provisioned PVCs that can be attached directly to executor pods
```

### Reuse PVC
This option allows reusing PVCs associated with Spark executors even after the executors are terminated (either due to consolidation activity or preemption in case of Spot instances).

This allows for preserving the intermediate shuffle data and cached data on the PVC. When Spark requests new executor pod to replace the terminated one, the system attempts to reuse an existing PVC that belonged to terminated executor. This option can be enabled by the following configuration:

`spark.kubernetes.executor.reusePersistentVolume=true`

> ### External Shuffle services like `Apache Celeborn`

Reviewer: ```suggestion
### External Shuffle services
```

> Leverage external shuffle service to decouple compute and storage, allowing spark executors to write data to external shuffle service instead of local disks, reducing the risk of data loss and data re-computation due to executor termination/consolidation.

Reviewer: ```suggestion
External shuffle services like Apache Celeborn decouple compute and storage, allowing Spark executors to write data to an external service instead of local disks. This reduces the risk of data loss and re-computation due to executor termination or consolidation.

```

> This allows for better resource management, when Spark Dynamic Resource Allocation is enabled.

Reviewer: Let's address why it's a good idea to use it with DRA

Also consider the performance implications of Apache Celeborn.
    * External shuffle services is typically suitable for extremely large volumes of shuffle data.

>     * For smaller datasets or applications with low shuffle data volues, the overhead of setting up and managing external shuffle service might outweigh its benefits.

Reviewer: fix indentation.

Is there a ball park number for `extremely large volumes`?
